larpandoracontent/LArVertex/VertexRefinementAlgorithm.cc

```diff
@@ -0,0 +1,250 @@
+        parameters.m_position = position3D;
+        parameters.m_vertexLabel = VERTEX_INTERACTION;
+        parameters.m_vertexType = VERTEX_3D;
+
+        const Vertex *pNewVertex(NULL);
+        PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::Vertex::Create(*this, parameters, pNewVertex));
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+CartesianVector VertexRefinementAlgorithm::RefineVertexTwoD(const ClusterList &clusterList, const CartesianVector &originalVtxPos) const
+{
+    CartesianPointVector intercepts, directions;
+    FloatVector weights;
+
+    for (const Cluster *const pCluster : clusterList)
+    {
+        if (LArClusterHelper::GetClosestDistance(originalVtxPos, pCluster) > 10)
+            continue;
+
+        if (pCluster->GetNCaloHits() < m_minimumHitsCut)
+            continue;
+
+        CartesianVector centroid(0.f, 0.f, 0.f);
+        LArPcaHelper::EigenValues eigenValues(0.f, 0.f, 0.f);
+        LArPcaHelper::EigenVectors eigenVectors;
+        CartesianPointVector pointVector;
+
+        LArClusterHelper::GetCoordinateVector(pCluster, pointVector);
+        LArPcaHelper::RunPca(pointVector, centroid, eigenValues, eigenVectors);
+
+        intercepts.push_back(LArClusterHelper::GetClosestPosition(originalVtxPos, pCluster));
+        directions.push_back(eigenVectors.at(0).GetUnitVector());
+        weights.push_back(1.f / ((LArClusterHelper::GetClosestPosition(originalVtxPos, pCluster) - originalVtxPos).GetMagnitudeSquared() + 1));
+    }
+
+    CartesianVector newVtxPos(originalVtxPos);
+    if (intercepts.size() > 1)
+        GetBestFitPoint(intercepts, directions, weights, newVtxPos);
+
+    if ((newVtxPos - originalVtxPos).GetMagnitude() > m_twoDDistanceCut)
+        return originalVtxPos;
+
+    return newVtxPos;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void VertexRefinementAlgorithm::GetBestFitPoint(const CartesianPointVector &intercepts, const CartesianPointVector &directions,
+    const FloatVector &weights, CartesianVector &bestFitPoint) const
+{
+    const int n(intercepts.size());
+
+    Eigen::VectorXd d = Eigen::VectorXd::Zero(3 * n);
+    Eigen::MatrixXd G = Eigen::MatrixXd::Zero(3 * n, n + 3);
+
+    for (int i = 0; i < n; ++i)
+    {
+        d(3 * i) = intercepts[i].GetX() * weights[i];
+        d(3 * i + 1) = intercepts[i].GetY() * weights[i];
+        d(3 * i + 2) = intercepts[i].GetZ() * weights[i];
+
+        G(3 * i, 0) = weights[i];
+        G(3 * i + 1, 1) = weights[i];
+        G(3 * i + 2, 2) = weights[i];
+
+        G(3 * i, i + 3) = -directions[i].GetX();
+        G(3 * i + 1, i + 3) = -directions[i].GetY();
+        G(3 * i + 2, i + 3) = -directions[i].GetZ();
+    }
+
+    if ((G.transpose() * G).determinant() == 0)
```

Comment:
I think we probably want to check that the absolute value of the determinant against a numeric_limits epsilon value here rather than identically zero.